### DISCUSSION TITLE
Use and care of the flatnode file

### DISCUSSION THREAD

**Discussion by akobelan** [osm-search/Nominatim#3265] [General]
I installed Nominatim several times at this point. As a learning experience to bail myself out when things go wrong. 

In one instance I forgot to specify using a flat node file. The question is, is there a way of generating the flatnode file post installation? 

And as a followup question, can I use the flatnode file from a previous installation by copying it to the appropriate location on a current installation?

**Comment by mtmail** [comment]
For data import and updates Nominatim uses osm2pgsql. osm2pgsql is a generic tool used by many OSM based projects. Nominatim's configuration is in the `settings/*.lua` files.

By default OSM data is written to the tables `planet_osm_rels`, `planet_osm_ways`, `planet_osm_nodes`. One can also use the flat node store (file). https://osm2pgsql.org/doc/manual.html#flat-node-store This store replaces the `planet_osm_nodes` file.

To recreate the file you'd need to run an import again with the same OSM input data into a new database. Then you could delete the database and copy the file. Folks in the osm2pgsql projects are more expert in this topic.

Afaik no other user asked about missing flatnode file, usually users start a new import from scratch.

> can I use the flatnode file from a previous installation by copying it to the appropriate location on a current installation

It has to match the database (the `planet_osm_rels`, `planet_osm_ways` tables). Copying from one host to another is possible, just like you'd dump and restore a database to another host.

**Comment by lonvia** [comment]
The flatnode file must match _exactly_ the data in the database or you will get very subtle errors. That means you can only copy the file from another server, if exactly the same import data was used and no updates have been run in the meantime.

A database loaded with a flatnode file also has a slightly different structure. I can't recommend adding it after the fact, even if there was a non-hacky way to produce one.

Note that you can duplicate a Nominatim installation by stopping updates, making a full DB dump, restoring the dump on the other machine, copying the flatnode file over, setting up the Nominatim configuration to point to the restored data and flatnode file and running `nominatim refresh --functions --website`.

**Reply by akobelan** [reply]
Then the question, "How can I tell if my current flatnode file is consistent with my database?", comes to mind. 

I'm running a Postgres instance on our network. And the "clients" are running on other machines not co-located with Postgres. Each machine would have it's own flatnode file. How does this work when upgrading the data in Postgres?

I see the usefulness of the flatnode file. I'm trying to understand how that comes into consideration for a single Postgres, multi Nominatim deployment.
